sklearn/pipeline.py

```diff
@@ -442,10 +573,63 @@ class FeatureUnion(BaseEstimator, TransformerMixin):
 
     """
     def __init__(self, transformer_list, n_jobs=1, transformer_weights=None):
-        self.transformer_list = transformer_list
+        self.transformer_list = tosequence(transformer_list)
         self.n_jobs = n_jobs
         self.transformer_weights = transformer_weights
+        self._validate_transformers()
 
+    def get_params(self, deep=True):
+        """Get parameters for this estimator.
+
+        Parameters
+        ----------
+        deep: boolean, optional
+            If True, will return the parameters for this estimator and
+            contained subobjects that are estimators.
+
+        Returns
+        -------
+        params : mapping of string to any
+            Parameter names mapped to their values.
+        """
+        return self._get_params('transformer_list', deep=deep)
+
+    def set_params(self, **kwargs):
+        """Set the parameters of this estimator.
+
+        Valid parameter keys can be listed with ``get_params``.
+
+        Returns
+        -------
+        self
+        """
+        self._set_params('transformer_list', **kwargs)
+        self._validate_transformers()
+        return self
+
+    def _validate_transformers(self):
+        names, transforms = zip(*self.transformer_list)
+
+        # validate names
+        self._validate_names(names)
+
+        # validate estimators
+        for t in transforms:
+            if t is None:
+                continue
+            if (not (hasattr(t, "fit") or hasattr(t, "fit_transform")) or not
+                    hasattr(t, "transform")):
+                raise TypeError("All estimators should implement fit and "
+                                "transform '%s' (type %s) doesn't)" %
+                                (t, type(t)))
+
+    def _iter(self):
+        get_weight = (self.transformer_weights or {}).get
+        return ((name, trans, get_weight(name))
+                for name, trans in self.transformer_list
+                if trans is not None)
+
+    @property
```

Comment:
I had meant that I was not convinced by relying on Python's implicit "try.. except Attribute Error" behviour for `hasattr` to return False when `get_features_name` is called. For instance, by chance if you access another attribute of the base transformer that is non-existent inside `get_features_name` of `FeatureUnion`, this will still return False.

That being said, this is the first time I've seen this being done and if you feel that this is common to do, go ahead and do it but do add a comment as to why this works.